src/bulk.rs

```diff
@@ -762,5 +764,770 @@ fn column_data_type_to_arrow(data_type: ColumnDataType) -> Result<DataType> {
+    builders: Vec<Box<dyn ArrayBuilder>>,
+    schema: Arc<Schema>,
+    current_rows: usize,
+}
+
+impl RowBatchBuilder {
+    /// Create a new RowBatchBuilder with the given schema and capacity
+    fn new(
+        table_schema: &[Column],
+        capacity: usize,
+        alloc_stats: Arc<AdaptiveAllocStats>,
+    ) -> Result<Self> {
+        let fields: Result<Vec<Field>> = table_schema
+            .iter()
+            .map(|col| {
+                column_data_type_to_arrow(col.data_type)
+                    .map(|data_type| Field::new(&col.name, data_type, true))
+            })
+            .collect();
+        let schema = Arc::new(Schema::new(fields?));
+
+        let builders: Result<Vec<Box<dyn ArrayBuilder>>> = table_schema
+            .iter()
+            .map(|col| create_array_builder_adaptive(col.data_type, capacity, alloc_stats.clone()))
+            .collect();
+
+        Ok(Self {
+            builders: builders?,
+            schema,
+            current_rows: 0,
+        })
+    }
+
+    /// Create a new RowBatchBuilder with a pre-computed Arrow schema
+    fn with_arrow_schema(
+        table_schema: &[Column],
+        schema: Arc<Schema>,
+        capacity: usize,
+        alloc_stats: Arc<AdaptiveAllocStats>,
+    ) -> Result<Self> {
+        let builders: Result<Vec<Box<dyn ArrayBuilder>>> = table_schema
+            .iter()
+            .map(|col| create_array_builder_adaptive(col.data_type, capacity, alloc_stats.clone()))
+            .collect();
+
+        Ok(Self {
+            builders: builders?,
+            schema,
+            current_rows: 0,
+        })
+    }
+
+    /// Add multiple rows to the batch builder using batch operations
+    fn add_rows(&mut self, rows: &[Row]) -> Result<()> {
+        for (col_idx, builder) in self.builders.iter_mut().enumerate() {
+            builder.append_values_from_rows(rows, col_idx)?;
+        }
+        self.current_rows += rows.len();
+        Ok(())
+    }
+
+    /// Build the RecordBatch from accumulated rows
+    fn build(mut self) -> Result<RecordBatch> {
+        let arrays: Result<Vec<Arc<dyn Array>>> = self
+            .builders
+            .iter_mut()
+            .map(|builder| builder.finish())
+            .collect();
+
+        let batch =
+            RecordBatch::try_new(self.schema, arrays?).context(error::CreateRecordBatchSnafu)?;
+
+        Ok(batch)
+    }
+
+    /// Get the current number of rows in the builder
+    fn len(&self) -> usize {
+        self.current_rows
+    }
+}
+
+/// Trait for type-erased array builders
+trait ArrayBuilder {
+    fn append_values_from_rows(&mut self, rows: &[Row], col_idx: usize) -> Result<()>;
+    fn finish(&mut self) -> Result<Arc<dyn Array>>;
+}
+
+/// Create an array builder for the given column data type with adaptive sizing
+fn create_array_builder_adaptive(
+    data_type: ColumnDataType,
+    capacity: usize,
+    alloc_stats: Arc<AdaptiveAllocStats>,
+) -> Result<Box<dyn ArrayBuilder>> {
```

Comment:
nit: 
If we want maximum performance, we can use enum to avoid the performance loss caused by dynamic dispatch.
And it is completely ok to keep the current code.

Reply:
Good idea, I'll do it.

Reply:
[069d7bb](https://github.com/GreptimeTeam/greptimedb-ingester-rust/pull/24/commits/069d7bb1e5824327e48d927074f36b7add396c0a)